src/Stratis.Bitcoin.Features.MemoryPool/Interfaces/ITxMempool.cs

```diff
@@ -215,14 +215,15 @@ public interface ITxMempool
         /// </summary>
         /// <param name="stage">Staged transactions.</param>
         /// <param name="updateDescendants">Whether to update decendants.</param>
+        /// <param name="removeForBlock">Whether the transaction is being removed because it was included in a block.</param>
         /// <remarks>
         /// If a transaction is in this set, then all in-mempool descendants must
         /// also be in the set, unless this transaction is being removed for being
         /// in a block.
         /// Set updateDescendants to true when removing a tx that was in a block, so
         /// that any in-mempool descendants have their ancestor state updated.
         /// </remarks>
-        void RemoveStaged(TxMempool.SetEntries stage, bool updateDescendants);
+        void RemoveStaged(TxMempool.SetEntries stage, bool updateDescendants, bool removeForBlock = false);
```

Comment:
We shouldn't need to track the reason why an unconfirmed transaction was removed from the mempool. It should be sufficient to know that a transaction is no longer unconfirmed.

Reply:
Why shouldn't we? 

There is a big difference between a transaction being actively "removed" from the mempool and a transaction being "confirmed" in a block. In the latter case, the removal from the mempool is a side effect of the chain advancing.

Reply:
Ideally it should be possible to accurately deduce the mempool/wallet state by simply looking at what is in the mempool and wallet (i.e. state driven) - without needing to know what other information may have flown between methods such as these. In that sense these changes should be a last resort and only be used if no better approach is available.